test/pullrequest/e2e_test.go

```diff
@@ -0,0 +1,170 @@
+		scmData string
+
+		// Auth token to use.
+		token string
+		// If auth token present, fmt template for Authorization header to check.
+		authHeader string
+	}{
+		{
+			name: "github-default",
+			// Must be http so we can intercept the request in the proxy.
+			url:        "http://github.com/wlynch/test/pulls/1",
+			scmData:    filepath.Join("testdata", "scm", "api.github.com"),
+			token:      envToken("GITHUB_TOKEN"),
+			authHeader: "Bearer %s",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			dir := filepath.Join(tmpdir, tc.name)
+			if err := os.Mkdir(dir, 0755); err != nil {
+				t.Fatalf("os.Mkdir: %v", err)
+			}
+
+			if tc.token != "" {
+				os.Setenv("AUTH_TOKEN", tc.token)
+				defer os.Unsetenv("AUTH_TOKEN")
+			}
+
+			if *proxy {
+				// Stand up fake proxy server. This should be removed if you want to talk to
+				// real SCM providers.
+				h := &handler{
+					Data: tc.scmData,
+					Header: http.Header{
+						"Authorization": []string{fmt.Sprintf(tc.authHeader, tc.token)},
+					},
+				}
+
+				l, err := net.Listen("tcp", "localhost:0")
+				if err != nil {
+					t.Fatalf("net.Listen: %v", err)
+				}
+				t.Logf("starting HTTP proxy on %s", l.Addr().String())
+				go func() {
+					if err := http.Serve(l, h); err != nil {
+						fmt.Println("http.Serve:", err)
+					}
+				}()
+				os.Setenv("http_proxy", l.Addr().String())
+			}
+
+			t.Log("Environment vars:", os.Environ())
+
+			banner(t, "download")
+
+			downloadFlags := []string{
+				"run",
+				"github.com/tektoncd/pipeline/cmd/pullrequest-init",
+				"-mode=download",
+				fmt.Sprintf("-path=%s", dir),
+				fmt.Sprintf("-url=%s", tc.url),
+			}
+			download := exec.Command("go", downloadFlags...)
+			t.Log(download)
+			if out, err := download.CombinedOutput(); err != nil {
+				t.Fatal(string(out))
+			}
+
+			golden := filepath.Join("testdata", "golden")
+			if diff, err := exec.Command("diff", "-r", golden, dir).CombinedOutput(); err != nil {
+				t.Error(string(diff))
+			}
+
+			// Modify PR resource contents
+			banner(t, "modify")
+			for path, v := range map[string]interface{}{
+				filepath.Join(dir, "status", "new.json"): &scm.Status{
+					State:  scm.StateSuccess,
+					Label:  "tekton-e2e",
+					Desc:   "tekton-e2e",
+					Target: "https://tekton.dev",
+				},
+			} {
+				f, err := os.Create(path)
+				if err != nil {
+					t.Fatalf("os.Create(%s): %v", path, err)
+				}
+				defer f.Close()
+				if err := json.NewEncoder(f).Encode(v); err != nil {
+					t.Fatalf("json.Encode(%s): %v", path, err)
+				}
+			}
+
+			// Upload new state back to server
+			banner(t, "upload")
+			uploadFlags := []string{
+				"run",
+				"github.com/tektoncd/pipeline/cmd/pullrequest-init",
+				"-mode=upload",
+				fmt.Sprintf("-path=%s", dir),
+				fmt.Sprintf("-url=%s", tc.url),
+			}
+			upload := exec.Command("go", uploadFlags...)
+			t.Log(upload)
+			out, err := upload.CombinedOutput()
+			if err != nil {
+				t.Fatal(string(out))
+			}
+		})
```

Comment:
i think this test would be a bit easier to read if some of this was broken into separate functions - probably exactly in the spots where you have comments to explain the code below, those are great candidates for separate functions

i dunno if youve ever read the book Clean Code (https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29#comments-rules has a good but probably too short summary of the contents) - ive only skimmed it myself!! but one thing that really stuck with me was to always try to explain myself in code and not comments - and often that means making a function call instead of a comment followed by a bunch of code

Reply:
nice!! i like it :D